rust/pytests/rootcanal.rs

```diff
@@ -0,0 +1,295 @@
+        } else {
+            None
+        };
+
+        child
+            .start_kill()
+            .unwrap_or_else(|e| warn!("Could not kill rootcanal: {:?}", e));
+        cancellation_token.cancel();
+        stdout_task
+            .await
+            .unwrap_or_else(|e| warn!("stdout task failed: {:?}", e));
+        stderr_task
+            .await
+            .unwrap_or_else(|e| warn!("stderr task failed: {:?}", e));
+        printer_task
+            .await
+            .unwrap_or_else(|e| warn!("print task failed: {:?}", e));
+
+        match child.wait().await {
+            Ok(exit) => debug!("exit status: {exit}"),
+            Err(e) => warn!("Error while waiting for child: {:?}", e),
+        }
+
+        if let Some(res) = test_res {
+            break res;
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct RootcanalPorts {
+    /// HCI TCP server port
+    pub(crate) hci_port: u16,
+    /// LE link TCP server port
+    pub(crate) link_ble_port: u16,
+    /// Link TCP server port
+    pub(crate) link_port: u16,
+    /// Test TCP port
+    pub(crate) test_port: u16,
+}
+
+/// Spawn a task that reads lines from `read` and writes them to `tx`
+fn spawn_output_copier<R: io::AsyncRead + Send + Unpin + 'static>(
+    read: R,
+    tx: mpsc::UnboundedSender<String>,
+) -> task::JoinHandle<()> {
+    tokio::spawn(async move {
+        let reader = io::BufReader::new(read);
+        let mut lines = reader.lines();
+
+        loop {
+            let res = lines.next_line().await;
+            match res {
+                Ok(None) => {
+                    // no more lines
+                    return;
+                }
+                Ok(Some(l)) => {
+                    if tx.send(l).is_err() {
+                        // rx closed
+                        return;
+                    };
+                }
+                Err(e) => {
+                    warn!("Could not read rootcanal output: {:?}", e)
+                }
+            }
+        }
+    })
+}
+
+/// Spawn a task to print output from rootcanal as it happens
+fn spawn_printer(
+    mut rx: mpsc::UnboundedReceiver<String>,
+    cancellation_token: CancellationToken,
+) -> task::JoinHandle<()> {
+    tokio::spawn(async move {
+        loop {
+            select! {
+                _ = cancellation_token.cancelled() => {
+                    // signal copier tasks to close
+                    rx.close();
+                    // print any buffered lines
+                    while let Some(line) = rx.recv().await {
+                        debug!("{line}")
+                    }
+
+                    break;
+                }
+                opt = rx.recv() => {
+                    match opt {
+                        None => break,
+                        Some(line) => debug!("{line}"),
+                    }
+                }
+            }
+        }
+    })
+}
+
+/// Returns the path to a rootcanal binary, downloading it from GitHub if necessary.
```

Comment:
It seems a little odd to me that this is downloading artifacts at runtime, would it make sense to move this step into a build.rs and setting this up at build time?

Reply:
I don't enjoy the current approach, but I don't see a great way of doing it with `build.rs` either that would work nicely with doing dirty tracking. I hope either way it's short lived, though -- with `n` additional hours I think I can get `rootcanal` building as a library, at which point a `rootcanal-sys` crate becomes reasonable, and this whole spawn-a-process thing can go away.